[Clang importer] Use ParseSourceFileRequest for parsing swift_attr attributes

DougGregor opened the pull request:
The Clang importer was directly calling into the parser to parse the
attribute (or modifier) within swift_attr. Aside from being gross, this
isn't possible with ASTGen.

Instead, teach ParseSourceFileRequest to deal with modifiers in the
same way that the Clang importer was hardcoding, and have the Clang
importer pull the attributes/modifiers off of the "missing"
declaration introduced by the request.

One benefit of this approach is that we're only parsing each
swift_attr source buffer once, then cloning the attributes each time
it's used, so we should be doing less work overall.

Fixes rdar://139119159.

DougGregor committed:
Introduce DeclAttribute::clone() to clone declaration attributes (obviously)

DougGregor committed:
[Clang importer] Use ParseSourceFileRequest for parsing swift_attr attributes

The Clang importer was directly calling into the parser to parse the
attribute (or modifier) within swift_attr. Aside from being gross, this
isn't possible with ASTGen.

Instead, teach ParseSourceFileRequest to deal with modifiers in the
same way that the Clang importer was hardcoding, and have the Clang
importer pull the attributes/modifiers off of the "missing"
declaration introduced by the request.

One benefit of this approach is that we're only parsing each
swift_attr source buffer once, then cloning the attributes each time
it's used, so we should be doing less work overall.

Fixes rdar://139119159.

DougGregor commented:
@swift-ci please smoke test

rintaro commented on include/swift/AST/Attr.h:
```diff
@@ -1802,6 +1902,11 @@ class CustomAttr final : public DeclAttribute {
 
+  /// Create a copy of this attribute.
+  CustomAttr *clone(ASTContext &ctx) const {
+    return create(ctx, AtLoc, getTypeExpr(), isImplicit());
```
`ArgumentList *argList;` and `PatternBindingInitializer *initContext;` are not cloned. Do we need to support `CustomAttr` in swift_attrs? If not, we should make this `UNIMPLEMENTED_CLONE`, then skip `CustomAttr` before cloning in ImportDecl (and diagnose).

DougGregor replied:
We do need `CustomAttr`, yes.

rintaro commented on include/swift/AST/Attr.h:
```diff
@@ -1131,6 +1163,8 @@ class TypeEraserAttr final : public DeclAttribute {
     return DA->getKind() == DeclAttrKind::TypeEraser;
   }
+
+  UNIMPLEMENTED_CLONE(DynamicReplacementAttr)
```
```suggestion
  UNIMPLEMENTED_CLONE(TypeEraserAttr)
```

DougGregor replied:
Nice catch, thank you!

rintaro reviewed (commented):
Thank you for doing this!

hamishknight reviewed (approved):
Nice!

DougGregor committed:
Add static checking to ensure every DeclAttribute subclass has a properly-typed clone()

Code review identified some incorrect UNIMPLEMENTED_CLONEs in DeclAttribute (thank you
Hamish and Rintaro). Fix those, and make sure this can't happen again by checking the type
signatures of clone() in every DeclAttribute subclass.

DougGregor committed:
Fix CustomAttr::clone() to include argument list and pattern binding initializer

DougGregor commented on include/swift/AST/Attr.h:
```diff
@@ -2639,6 +2812,8 @@ class RawLayoutAttr final : public DeclAttribute {
     return DA->getKind() == DeclAttrKind::RawLayout;
   }
+
+  UNIMPLEMENTED_CLONE(MacroRoleAttr)
```
Thank you! I've added some static checking so this can't happen again.

DougGregor commented on lib/Parse/ParseRequests.cpp:
```diff
@@ -379,10 +379,12 @@ SourceFileParsingResult parseSourceFile(SourceFile &SF) {
+      parser.parseExpandedAttributeList(items, /*isFromClangAttribute=*/false);
+      break;
+
+    case GeneratedSourceInfo::Attribute:
```
Yes, that's a good idea.

DougGregor committed:
Renable GeneratedSourceInfo::Attribute to GeneratedSourceInfo::AttributeFromClang

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test windows
